pkg/retrieval/retrieval.go

```diff
@@ -112,134 +112,88 @@ func (s *Service) RetrieveChunk(ctx context.Context, addr swarm.Address, origin
 		flightRoute = addr.String() + originSuffix
 	}
 
-	// topCtx is passing the tracing span to the first singleflight call
-	topCtx := ctx
-
-	v, _, err := s.singleflight.Do(ctx, flightRoute, func(ctx context.Context) (interface{}, error) {
-		maxPeers := 1
-		if origin {
-			maxPeers = maxSelects
-		}
+	v, _, err := s.singleflight.Do(ctx, flightRoute, func(_ context.Context) (interface{}, error) {
 
 		sp := newSkipPeers()
 
-		ticker := time.NewTicker(retrieveRetryIntervalDuration)
-		defer ticker.Stop()
-
 		var (
-			peerAttempt  int
-			peersResults int
-			resultC      = make(chan retrievalResult, maxSelects)
+			allowedRetries   = maxRetries
+			allowedRetrieves = 1
 		)
 
-		requestAttempt := 0
+		if origin {
+			allowedRetrieves = maxRetrieved
+		}
+
+		timer := time.NewTimer(0)
+		defer timer.Stop()
 
-		lastTime := time.Now().Unix()
+		resultChan := make(chan retrievalResult)
+		doneChan := make(chan struct{})
+		defer close(doneChan)
 
-		for requestAttempt < maxRequestRounds {
+		for {
 
-			if peerAttempt < maxSelects {
+			select {
+			case <-ctx.Done():
+				return nil, storage.ErrNotFound
+			case <-timer.C:
+
+				allowedRetries--
+				allowedRetrieves--
+
+				s.metrics.PeerRequestCounter.Inc()
 
 				// create a new context without cancelation but
 				// set the tracing span to the new context from the context of the first caller
-				ctx := tracing.WithContext(context.Background(), tracing.FromContext(topCtx))
+				ctx := tracing.WithContext(context.Background(), tracing.FromContext(ctx))
 
 				// get the tracing span
 				span, _, ctx := s.tracer.StartSpanFromContext(ctx, "retrieve-chunk", s.logger, opentracing.Tag{Key: "address", Value: addr.String()})
 				defer span.Finish()
 
-				peerAttempt++
-				s.metrics.PeerRequestCounter.Inc()
 				go func() {
-
 					// cancel the goroutine just with the timeout
-					ctx, cancel := context.WithTimeout(ctx, retrieveChunkTimeout)
+					ctx, cancel := context.WithTimeout(ctx, defaultTTL)
 					defer cancel()
-
-					chunk, peer, requested, err := s.retrieveChunk(ctx, addr, sp, origin)
-					select {
-					case resultC <- retrievalResult{
-						chunk:     chunk,
-						peer:      peer,
-						err:       err,
-						retrieved: requested,
-					}:
-					case <-ctx.Done():
-					}
-
+					s.retrieveChunk(ctx, resultChan, doneChan, addr, sp, origin)
 				}()
-			} else {
-				select {
-				case resultC <- retrievalResult{}:
-				case <-ctx.Done():
+
+				if allowedRetries <= 0 || allowedRetrieves <= 0 {
```

Comment:
this looks a bit odd... if we have no more attempts left then we try to do another attempt?